modules/generator/remotewrite/appender.go

```diff
@@ -0,0 +1,156 @@
+package remotewrite
+
+import (
+	"context"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/klauspost/compress/snappy"
+	"github.com/prometheus/prometheus/model/exemplar"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/prometheus/prometheus/storage"
+)
+
+// TODO: make this configurable
+var maxWriteRequestSize = 3 * 1024 * 1024 // 3MB
+
+// remoteWriteAppender is a storage.Appender that remote writes samples and exemplars.
+type remoteWriteAppender struct {
+	logger       log.Logger
+	ctx          context.Context
+	remoteWriter *remoteWriteClient
+	userID       string
+
+	// TODO Loki uses util.EvictingQueue here to limit the amount of samples written per remote write request
+	labels         [][]prompb.Label
+	samples        []prompb.Sample
+	exemplarLabels [][]prompb.Label
+	exemplars      []prompb.Exemplar
+
+	metrics *Metrics
+}
+
+var _ storage.Appender = (*remoteWriteAppender)(nil)
+
+func (a *remoteWriteAppender) Append(_ storage.SeriesRef, l labels.Labels, t int64, v float64) (storage.SeriesRef, error) {
+	a.labels = append(a.labels, labelsToLabelsProto(l))
+	a.samples = append(a.samples, prompb.Sample{
+		Timestamp: t,
+		Value:     v,
+	})
+	return 0, nil
+}
+
+func (a *remoteWriteAppender) AppendExemplar(_ storage.SeriesRef, l labels.Labels, e exemplar.Exemplar) (storage.SeriesRef, error) {
+	a.exemplarLabels = append(a.exemplarLabels, labelsToLabelsProto(l))
+	a.exemplars = append(a.exemplars, prompb.Exemplar{
+		Labels:    labelsToLabelsProto(e.Labels),
+		Value:     e.Value,
+		Timestamp: e.Ts,
+	})
+	return 0, nil
+}
+
+func (a *remoteWriteAppender) Commit() error {
+	level.Debug(a.logger).Log("msg", "writing samples to remote_write target", "tenant", a.userID, "target", a.remoteWriter.Endpoint(), "count", len(a.samples))
```

Comment:
nit: Not sure if we should log `.Endpoint()` here as it can contain user/pass credentials? Is this true?

Thinking if we should use `.Name()` but I saw you're using the same name for all clients `metrics_generator`. I don't think we can keep the name unique because it seems to be used as a label in some metrics.

Same comment to the `.Endpoint()` logging done a few lines below.

Reply:
Good spotting, I'm not entirely sure tbh 🤔  Looking at the Prometheus code we use, this _seems_ fine.

I think it's always possible that someone configures a URL like `http://<username>:<password>@example.com/`, which would get logged in this case.

This is not really how it is intended to be used though: the remote write config struct has dedicated fields to configure your credentials. `metrics_generator.remote_write.client` is just this Prometheus struct: https://prometheus.io/docs/prometheus/latest/configuration/configuration/#remote_write
So there are fields like `basic_auth`, `authorization` and `oauth2` to configure credentials.

We pass this config to a Prometheus `WriteClient`. The `WriteClient` interface only exposes `Name` and `Endpoint` which makes it seem like this would be safe to print: https://github.com/prometheus/prometheus/blob/e239e3ee8b13b51b0f791a199813a14f74600a7e/storage/remote/queue_manager.go#L327-L336

Another data point: when sending the request fails, this is logged:
```
level=error ts=2022-02-17T14:28:59.380752142Z caller=appender.go:70 msg="error sending remote-write requests" tenant=116095 target=https://xxx.grafana.net/api/prom/push err="Post \"https://xxx.grafana.net/api/prom/push\": EOF"
```
Note that the error that is returned by the write client also contains the full endpoint.

Reply:
Tracked this down. No credentials should be published in the metric label: https://github.com/prometheus/prometheus/pull/9383

Reply:
Nice find!